src/file.cpp

```diff
@@ -46,6 +46,147 @@ File::~File()
     }
 }
 
+bool File::serialize(string *d)
+{
+    char type = transfer->type;
+    d->append((const char*)&type, sizeof(type));
+
+    if (!FileFingerprint::serialize(d))
+    {
+        LOG_err << "Error serializing File: Unable to serialize FileFingerprint";
+        return false;
+    }
+
+    unsigned short ll;
+    bool flag;
+
+    ll = name.size();
+    d->append((char*)&ll, sizeof(ll));
+    d->append(name.data(), ll);
+
+    ll = localname.size();
+    d->append((char*)&ll, sizeof(ll));
+    d->append(localname.data(), ll);
+
+    ll = privauth.size();
+    d->append((char*)&ll, sizeof(ll));
+    d->append(privauth.data(), ll);
+
+    ll = pubauth.size();
+    d->append((char*)&ll, sizeof(ll));
+    d->append(pubauth.data(), ll);
+
+    d->append((const char*)&h, sizeof(h));
+    d->append((const char*)filekey, sizeof(filekey));
+
+    flag = hprivate;
+    d->append((const char*)&flag, sizeof(flag));
+
+    flag = hforeign;
+    d->append((const char*)&flag, sizeof(flag));
+
+    flag = syncxfer;
+    d->append((const char*)&flag, sizeof(flag));
+
+    return true;
+}
+
+File *File::unserialize(string *d)
+{
+    if (!d->size())
+    {
+        LOG_err << "Error unserializing File: Empty string";
+        return NULL;
+    }
+
+    d->erase(0, 1);
+
+    FileFingerprint *fp = FileFingerprint::unserialize(d);
+    if (!fp)
+    {
+        LOG_err << "Error unserializing File: Unable to unserialize FileFingerprint";
+        return NULL;
+    }
+
+    const char* ptr = d->data();
+    const char* end = ptr + d->size();
+
+    if (ptr + sizeof(unsigned short) > end)
+    {
+        LOG_err << "File unserialization failed - serialized string too short";
+        return NULL;
+    }
+
+    // read name
+    unsigned short namelen = MemAccess::get<unsigned short>(ptr);
+    ptr += sizeof(namelen);
+    if (ptr + namelen + sizeof(unsigned short) > end)
```

Comment:
That `sizeof(unsigned short)` is to ensure that the next `unsigned short` that is going to be read (`localnamelen`, few lines below) is available.